### PR TITLE
 RUN-2846 Send licensing info

### PR DESCRIPTION
**Pull request by datamadic**
This front-runs the rvm work. The basic logic is a follows: 

If you have a license key in your manifest, send it along on app run
If you have a license key in your app options, use that, else use the manifest key
If you do not have an options or manifest license key, use your parent's (if any)

Many of the comments that were for the [first iteration of this PR](https://github.com/HadoukenIO/core/pull/41) have been addressed.

[tests well](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59381035998bc04e20cc6cc4) off a fresh merge to develop

### CODE REVIEW

File: src/browser/core_state.js [hunk 1]
```diff
@@ -512,6 +514,30 @@ function getAppAncestor(descendantAppUuid) {
...
+function setLicenseKey(identity, licenseKey) {
+    const { uuid } = identity;
+    const app = getAppByUuid(uuid);
+
+    if (app) {
+        app.licenseKey = licenseKey;
+
+        return licenseKey;
+    } else {
+        return null;
```
Comment (HarsimranSingh) [inline src/browser/core_state.js]: Why not return `undefined`?

File: src/browser/core_state.js [hunk 2]
```diff
@@ -512,6 +514,30 @@ function getAppAncestor(descendantAppUuid) {
...
+}
+
+function getLicenseKey(identity) {
+    const { uuid } = identity;
+    const app = getAppByUuid(uuid);
+
+    if (app) {
+        return app.licenseKey;
+    } else {
+        return null;
```
Comment (HarsimranSingh) [inline src/browser/core_state.js]: Same here

File: src/browser/core_state.js (hunk 1 above)
Comment (HarsimranSingh) [inline src/browser/core_state.js]: Actually, why does this return anything in the first place?

File: src/browser/core_state.js [hunk 3]
```diff
@@ -124,6 +124,8 @@ function appByUuid(uuid) {
     return coreState.apps.find(app => uuid === app.uuid);
 }
 
+const getAppByUuid = appByUuid;
```
Comment (datamadic) [inline src/browser/core_state.js]: So I spent an embarrassing amount of time looking for how to get an app by uuid. All the other getters start with `get` so this is just for consistency: 

````js
appByUuid,
// ...
// ...
getAppById,
getAppByWin,
getAppObj,
getAppObjByUuid,
````

File: src/browser/core_state.js (hunk 1 above)
Comment (datamadic) [inline src/browser/core_state.js]: Just a deterministic return as opposed to `undefined`  This also gives you an easy way to know if you actually set it w/o having to `throw`

File: src/browser/core_state.js (hunk 2 above)
Comment (datamadic) [inline src/browser/core_state.js]: `see above`

File: index.js [hunk 4]
```diff
@@ -481,11 +485,13 @@ function launchApp(argo, startExternalAdapterServer) {
 }
 
 
-function initFirstApp(options, configUrl) {
+function initFirstApp(options, configUrl, licenseKey) {
     try {
         // Needs proper configs
         firstApp = Application.create(options, configUrl);
 
+        coreState.setLicenseKey({ uuid: options.uuid }, licenseKey);
```
Comment (datamadic) [inline index.js]: where is the other one?

Comment (datamadic) [inline index.js]: yes, sorry, will fix

File: src/browser/core_state.js (hunk 3 above)
Comment (HarsimranSingh) [inline src/browser/core_state.js]: Discussed offline, we should eventually switch over all usages of `appByUuid` to `getAppByUuid` to better conform to the convention.

File: src/browser/core_state.js (hunk 1 above)
Comment (HarsimranSingh) [inline src/browser/core_state.js]: 👍

File: src/browser/core_state.js (hunk 2 above)
Comment (HarsimranSingh) [inline src/browser/core_state.js]: 👍

**Comment by HarsimranSingh** [comment]
👍

File: test/rvm_message_bus.test.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import * as assert from 'assert';
+import * as mockery from 'mockery';
+import {EventEmitter} from 'events';
+const App = require('electron').app;
```
Comment (whyn07m3) [inline test/rvm_message_bus.test.ts]: `import { app } from 'electron';`

File: test/rvm_message_bus.test.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                    version: null,
+                    pid: null
+                }
+            }
+            const {payload} = <any> rvmMessageBus.registerLicenseInfo(<LicenseInfo> {
+                processId: 'processId',
+                runtimeVersion: 'runtimeVersion'
+            });
+
+            assert.deepEqual(payloadShape, payload, 'shapes should match hommie');
```
Comment (whyn07m3) [inline test/rvm_message_bus.test.ts]: Let's keep it professional :bowtie:

File: test/rvm_message_bus.test.ts [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+limitations under the License.
+*/
+import * as assert from 'assert';
+import * as mockery from 'mockery';
+import {EventEmitter} from 'events';
+const App = require('electron').app;
+
+App.generateGUID = Math.random;
+
+const mockWMCopyData  = {
```
Comment (whyn07m3) [inline test/rvm_message_bus.test.ts]: I had to (mostly spaces) :)
```
const mockWMCopyData = {
   WMCopyData: () => {
      return {
         on: (x: any) => x,
         publish: (x: any) => x
      }
   }
}
```

File: test/rvm_message_bus.test.ts (hunk 5 above)
Comment (datamadic) [inline test/rvm_message_bus.test.ts]: this avoids the ts compiler from telling us that `generateGUID ` does not exist on electron's `app`. It is something that we added

File: test/rvm_message_bus.test.ts (hunk 6 above)
Comment (datamadic) [inline test/rvm_message_bus.test.ts]: yes, will fix, sorry this somehow found its way back in...

File: test/rvm_message_bus.test.ts (hunk 7 above)
Comment (datamadic) [inline test/rvm_message_bus.test.ts]: yes, will fix

File: test/rvm_message_bus.test.ts (hunk 5 above)
Comment (whyn07m3) [inline test/rvm_message_bus.test.ts]: Can you add `generateGUID` [here](https://github.com/HadoukenIO/core/blob/develop/src/modules.d.ts#L23) to solve this issue?

Comment (datamadic) [inline test/rvm_message_bus.test.ts]: learn something new everyday.. I will add and retest

File: test/rvm_message_bus.test.ts (hunk 7 above)
Comment (datamadic) [inline test/rvm_message_bus.test.ts]: so this is not treated as a "constructor" in `=>` syntax 😄 I'll leave it as is

**Comment by datamadic** [comment]
[updated tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59384faa998bc04e20cc6cd9)

**Pull request merged**